### DISCUSSION TITLE
list_cumulative: window function equivalent of list_aggregate

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#11541] [Ideas]
It would be very convenient if there was a "cumulative window" equivalent of `list_aggregate`.

This `list_cumulative` function, would apply arbitrary window functions to a column of `list` type with a implicit `ORDER BY list_index ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW` framing within individual `list` entries of the column.

For example,
```sql
SELECT 
  -- windowed aggregate function
  list_cumulative([3, 4, 6], 'sum') AS cumulativeSum, -- [3, 7, 13]
  list_cumulative([3, 4, 6], 'median') AS runningMedian, -- [3, 3.5, 4]
  -- general purpose window functions
  list_cumulative([3, NULL, 6], 'last', ignore_nulls := true) AS fillforward, -- [3, 3, 6]
  list_zip(list_cumulative(['a', 'b', 'c'], 'row_number'), ['a', 'b', 'c']) AS enumerate, -- [(1, 'a'), (2, 'b'), (3, 'c')]
  unnest(list_cumulative(['a', 'b', 'c'], 'row_number')) AS i, -- emulation of duckdb's generate_subscripts
```

Just as for `list_aggregate` functions, there could be predefined rewrites for even more convenience: `list_cumulative_sum`, `list_cumulative_median`, `list_cumulative_max`, etc. 

Implementing `list_cumulative` manually is very tedious, not composable, and full of traps.
For example, 
```sql
SELECT
  list_cumulative(a, 'median') AS runningMedian
FROM (VALUES ([1, 2], 'a'), ([3, 4, 5], 'b'), ([1, 2], 'a'), ([1, 2], 'c')) _(a, b)
```
```
┌───────────┬─────────┬─────────────────┐
│     a     │    b    │  runningMedian  │
│  int32[]  │ varchar │    double[]     │
├───────────┼─────────┼─────────────────┤
│ [3, 4, 6] │ b       │ [3.0, 3.5, 4.0] │
│ [1, 2]    │ a       │ [1.0, 1.5]      │
│ [1, 2]    │ a       │ [1.0, 1.5]      │
│ [1, 2]    │ c       │ [1.0, 1.5]      │
└───────────┴─────────┴─────────────────┘
```
is equivalent to 
```sql
SELECT
  first(a) AS a, 
  first(b) AS b,
  list(runningMedian ORDER BY _i) AS runningMedian
FROM
(
    SELECT
      *, 
      _i,
      median(unnested) OVER (PARTITION BY _a ORDER BY _i) as runningMedian,
    FROM
    (
        SELECT 
          *,
          row_number() OVER () AS _a,
          generate_subscripts(a, 1) AS _i,
          a[_i] AS unnested
        FROM (VALUES ([1, 2], 'a'), ([3, 4, 5], 'b'), ([1, 2], 'a'), ([1, 2], 'c')) _(a, b)
    )
)
GROUP BY (_a)
```

Even for the most simple window function, the cumulative sum, although there is a shorter solution
```sql
SELECT 
  list_reduce(list_transform(l, x -> list_value(x)), (x, y) -> list_append(x, x[-1] + y[1]))
FROM df
```
it's very ugly and requires O(n^2) (slightly higher exponent even in my experiments).

---

**Why not a general window instead of just cumulative?**

Note that the lack of partitioning support in `list_cumulative_window` isn't a big loss in the list column context: within a list, you can't partition by anything other than the entries of the list itself anyway, so you're only losing out on exotic things like `"list_window([1, 1, 3], 'sum' PARTITION BY list_value)" = [2, 2, 3]`. 

Similarly, the lack of order and framing support isn't a big loss, because the only other framing you could define within a `ORDER BY list_index` list is `ROWS BETWEEN integerconstant PRECEDING AND integerconstant FOLLOWING`, so you're missing out on `"list_window([1, 2, 3], 'sum' RANGE BETWEEN 1 PRECEDING AND CURRENT ROW)" = [1, 3, 5]`. Finally, the only other order you could logically define within a list is by the list value, but if you want that you could just sort your list first (except if you want `ORDER BY value RANGE BETWEEN ...` framing).

---

In summary, I think a simple `list_cumulative` gets 90% of what one could want in terms of window functions for a list column without the questionable syntax hacks required for the exotic examples above, and it saves 20 rows of tricky SQL syntax.
